stix2/utils.py

```diff
@@ -29,32 +48,47 @@ def format_datetime(dttm):
     else:
         zoned = dttm.astimezone(pytz.utc)
     ts = zoned.strftime("%Y-%m-%dT%H:%M:%S")
-    if zoned.microsecond > 0:
-        ms = zoned.strftime("%f")
+    ms = zoned.strftime("%f")
+    precision = getattr(dttm, "precision", None)
+    if precision:
+        if precision == "millisecond":
+            ts = ts + '.' + ms[:3]
+    elif zoned.microsecond > 0:
         ts = ts + '.' + ms.rstrip("0")
     return ts + "Z"
 
 
-def parse_into_datetime(value):
+def parse_into_datetime(value, precision=None):
```

Comment:
I'd feel a bit better if we had some more tests to handle the additional complexity of this function.